docs/proposals/20190919-machinepool-api.md

```diff
@@ -0,0 +1,361 @@
+    - Type: `*int32`
+    - Description: Minimum number of seconds for which a newly created machine should be ready.
+  - **ProviderIDs [optional]**
+    - Type: `[]Instance`
+    - Description: ProviderIDs contain a ProviderID for each machine instance that's currently
+      managed by the infrastructure provider belonging to the machine pool.
+
+``` go
+type MachinePoolStatus struct
+```
+
+- **To add**
+  - **NodeRefs [optional]**
+    - Type: `[]corev1.ObjectReference`
+    - Description: NodeRefs contain a NodeRef for each ProviderID in MachinePoolSpec.ProviderIDs.
+  - **Replicas [optional]**
+    - Type: `*int32`
+    - Description: Replicas is the most recent observed number of replicas.
+  - **ReadyReplicas [optional]**
+    - Type: `*int32`
+    - Description: The number of ready replicas (ready for at least minReadySeconds) for this
+      MachinePool.
+  - **AvailableReplicas [optional]**
+    - Type: `*int32`
+    - Description: The number of available replicas (ready for at least minReadySeconds) for this
+      MachinePool.
+  - **UnavailableReplicas [optional]**
+    - Type: `*int32`
+    - Description: Total number of unavailable machines targeted by this machine pool. This is the
+      total number of machines that are still required for this machine pool to have 100% available
+      capacity. They may either be machines that are running but not yet available or machines that
+      still have not been created.
+  - **ErrorReason [optional]**
+    - Type: `*capierrors.MachinePoolStatusError`
+    - Description: ErrorReason indicates that there is a problem reconciling the state, and will be
+      set to a token value suitable for programmatic interpretation.
+  - **ErrorMessage [optional]**
+    - Type: `*string`
+    - Description: ErrorMessage indicates that there is a problem reconciling the state, and will be
+      set to a descriptive error message.
+  - **Phase [optional]**
+    - Type: `string`
+	- Description: Phase represents the current phase of cluster actuation.
+	  e.g. Pending, Running, Terminating, Failed etc.
+  - **BootstrapReady [optional]**
+    - Type: `bool`
+    - Description: True when the bootstrap provider status is ready.
+  - **InfrastructureReady [optional]**
+    - Type: `bool`
+    - Description: True when the infrastructure provider status is ready.
+
+#### States and Transitions
+
+##### Pending
+
+``` go
+// MachinePoolPhasePending is the first state a MachinePool is assigned by
+// Cluster API MachinePool controller after being created.
+MachinePoolPhasePending = MachinePoolPhase("pending")
+```
+
+###### Transition Conditions
+
+- MachinePool.Phase is empty
+
+###### Expectations
+
+- When MachinePool.Spec.Template.Spec.Bootstrap.Data is:
+  - <nil>, expect the field to be set by an external controller.
+  - “” (empty string), expect the bootstrap step to be ignored.
+  - “...” (populated by user or from the bootstrap provider), expect the contents to be used by a
```

Comment:
Capturing @ncdc's comment from the google doc:

In the case of CABPK, does it need to be modified to generate bootstrap data for KubeadmConfigs that are owned by a MachinePool? Or what triggers CABPK to generate bootstrap data for the MachinePool?

Reply:
@juan-lee checking in on this comment - do you think you'll be able to address it soon? Thanks!

Reply:
CABPK needs to be modified to know about about MachinePool. The interaction is nearly identical compared to Machine, except for control plane concerns.

For reference:
https://github.com/juan-lee/cluster-api/blob/machinepool-caep/bootstrap/kubeadm/controllers/kubeadmconfig_controller.go#L122

The code can be improved, but it demonstrates how CABPK interacts with MachinePool.

Reply:
@ncdc @vincepri  I filed issue #1799 to track a better approach than in my POC.